p4include/psa.p4

```diff
@@ -0,0 +1,675 @@
+    void reset        (in S index);
+    void start        (in S index);
+    void stop         (in S index);
+  }
+  */
+}
+
+/// DirectCounter: RFC-XXXX
+extern DirectCounter<W> {
+  DirectCounter(entry_key key, CounterType_t counter_type);
+  void count();
+
+  /*
+  @ControlPlaneAPI
+  {
+    W    read<W>      (in entry_key key);
+    W    sync_read<W> (in entry_key key);
+    void set          (in W seed);
+    void reset        (in entry_key key);
+    void start        (in entry_key key);
+    void stop         (in entry_key key);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Meters: RFC 2698
+ *
+ * Meters are a more complex mechanism for keeping statistics about
+ * the packets that trigger a table. The meters specified in the PSA
+ * are 3-color meters.
+ */
+
+/// Meter types
+enum MeterType_t {
+    packets,
+    bytes
+}
+/// Meter colors
+enum MeterColor_t { RED, GREEN, YELLOW };
+
+/// Meter
+extern Meter<S> {
+  Meter(S n_meters, MeterType_t type);
+  MeterColor_t execute(in S index, in MeterColor_t color);
+
+  /*
+  @ControlPlaneAPI
+  {
+    reset(in MeterColor_t color);
+    setParams(in S committedRate, out S committedBurstSize
+              in S peakRate, in S peakBurstSize);
+    getParams(out S committedRate, out S committedBurstSize
+              out S peakRate, out S peakBurstSize);
+  }
+  */
+}
+
+/// DirectMeter
+extern DirectMeter {
+  DirectMeter(MeterType_t type);
+  MeterColor_t execute(in MeterColor_t color);
+
+  /*
+  @ControlPlaneAPI
+  {
+    reset(in MeterColor_t color);
+    void setParams<S>(in S committedRate, out S committedBurstSize
+                      in S peakRate, in S peakBurstSize);
+    void getParams<S>(out S committedRate, out S committedBurstSize
+                      out S peakRate, out S peakBurstSize);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Registers
+ *
+ * Registers are stateful memories whose values can be read and
+ * written in actions. Registers are similar to counters, but can be
+ * used in a more general way to keep state.
+ *
+ * Although registers cannot be used directly in matching, they may be
+ * used as the RHS of an assignment operation, allowing the current
+ * value of the register to be copied into metadata and be available
+ * for matching in subsquent tables.
+ */
+extern Register<T, S> {
+  Register(S size);
+  T    read  (in S index);
+  void write (in S index, in T value);
+
+  /*
+  @ControlPlaneAPI
+  {
+    T    read<T>      (in S index);
+    void set          (in S index, in T seed);
+    void reset        (in S index);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Random
+ *
+ * The random extern provides a reliable, target specific number generator
+ * in the min .. max range.
+ */
+
+/// The set of distributions supported by the Random extern.
+enum RandomDistribution {
+  PRNG,
```

Comment:
What do you think about specifying the distribution in general? It is likely that most hardware implementation of the Random extern will implement a fixed algorithm, and it is not clear how do we map all the PSA distribution to that one implementation. A software implementation could call different libraries and get the random distribution. So the question is: is it more portable to just say a random number generator with whatever algorithm is implemented in the target?